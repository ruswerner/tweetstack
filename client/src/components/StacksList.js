import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import Stack from './Stack';
import StackRecord from '../records/stack';

export default class StacksList extends PureComponent {

  static propTypes = {
    stacks: ImmutablePropTypes.listOf(StackRecord).isRequired,
    onRefresh: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired,
    onSearch: PropTypes.func.isRequired
  };

  constructor() {
    super();
    this.scroll = null;
  }

  componentDidUpdate(prevProps) {
    // If a new stack was added and there are at least 2 stacks, auto-scroll the new stack into view
    if (prevProps.stacks.size < this.props.stacks.size && this.props.stacks.size > 1) {
      this.scroll.scrollLeft = this.scroll.scrollWidth;
    }
  }

  componentWillUnmount() {
    this.scroll = null;
  }

  render() {

    const justifyContent = this.props.stacks.size === 1 ?
      'center' :
      'flex-start';

    return (
      <div style={{overflow: 'scroll'}}
           ref={n => this.scroll = n}>
        <div style={{
          padding: 15,
          paddingBottom: 30,
          display: 'flex',
          flexDirection: 'row',
          justifyContent
        }}>
          {this.props.stacks.map((stack, i) =>
            <Stack
              key={`${stack.query}${i}`}
              id={`stack${i}`}
              onRefresh={() => this.props.onRefresh(i)}
              onRemove={() => this.props.onRemove(i)}
              onSearch={q => this.props.onSearch(i, q)}
              query={stack.query}
              tweets={stack.tweets}
              loading={stack.loading}
              noResults={stack.noResults}
              error={stack.error}
            />)
          }
        </div>
      </div>
    );
  }

}
