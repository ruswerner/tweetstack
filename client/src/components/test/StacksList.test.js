import React from 'react';
import Immutable from 'immutable';
import renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import StacksList from '../StacksList';
import StackRecord from '../../records/stack';

const noop = () => null;

describe('StacksList', function() {

  describe('when no stacks', function() {
    it('should match snapshot', function() {
      const tree = renderer.create(
        <MuiThemeProvider>
          <StacksList
            stacks={new Immutable.List()}
            onRefresh={noop}
            onRemove={noop}
            onSearch={noop}
          />
        </MuiThemeProvider>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when 1 stack', function() {
    it('should match snapshot', function() {
      const tree = renderer.create(
        <MuiThemeProvider>
          <StacksList
            stacks={new Immutable.List([new StackRecord()])}
            onRefresh={noop}
            onRemove={noop}
            onSearch={noop}
          />
        </MuiThemeProvider>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

});
