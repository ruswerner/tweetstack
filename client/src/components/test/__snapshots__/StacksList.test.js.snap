// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StacksList when 1 stack should match snapshot 1`] = `
<div
  style={
    Object {
      "overflow": "scroll",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
        "padding": 15,
        "paddingBottom": 30,
      }
    }
  >
    <div>
      <div
        style={
          Object {
            "WebkitTapHighlightColor": "rgba(0,0,0,0)",
            "backgroundColor": "#ffffff",
            "borderRadius": 2,
            "boxShadow": "0 10px 30px rgba(0, 0, 0, 0.19),
                   0 6px 10px rgba(0, 0, 0, 0.23)",
            "boxSizing": "border-box",
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "flex",
            "flexDirection": "column",
            "fontFamily": "Roboto, sans-serif",
            "marginRight": 15,
            "maxWidth": 400,
            "minHeight": 300,
            "minWidth": 400,
            "muiPrepared": true,
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          }
        }
      >
        <div
          className={undefined}
          style={
            Object {
              "WebkitTapHighlightColor": "rgba(0,0,0,0)",
              "backgroundColor": "rgb(232, 232, 232)",
              "boxSizing": "border-box",
              "display": "flex",
              "height": 56,
              "justifyContent": "space-between",
              "muiPrepared": true,
              "padding": "0px 24px",
              "paddingRight": 0,
            }
          }
        >
          <div
            className={undefined}
            style={
              Object {
                "backgroundColor": "transparent",
                "cursor": "auto",
                "display": "inline-block",
                "fontFamily": "Roboto, sans-serif",
                "fontSize": 16,
                "height": 48,
                "lineHeight": "24px",
                "muiPrepared": true,
                "position": "relative",
                "transition": "height 200ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                "width": 256,
              }
            }
          >
            <div
              style={
                Object {
                  "bottom": 12,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "muiPrepared": true,
                  "opacity": 1,
                  "position": "absolute",
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                }
              }
            >
              Search...
            </div>
            <input
              disabled={false}
              id="stack0"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              style={
                Object {
                  "WebkitOpacity": 1,
                  "WebkitTapHighlightColor": "rgba(0,0,0,0)",
                  "appearance": "textfield",
                  "backgroundColor": "rgba(0,0,0,0)",
                  "border": "none",
                  "color": "rgba(0, 0, 0, 0.87)",
                  "cursor": "inherit",
                  "font": "inherit",
                  "height": "100%",
                  "muiPrepared": true,
                  "outline": "none",
                  "padding": 0,
                  "position": "relative",
                  "width": "100%",
                }
              }
              type="text"
              value=""
            />
            <div>
              <hr
                aria-hidden="true"
                style={
                  Object {
                    "borderBottomStyle": "solid",
                    "borderBottomWidth": 1,
                    "borderColor": "#e0e0e0",
                    "borderLeft": "none",
                    "borderRight": "none",
                    "borderTop": "none",
                    "bottom": 8,
                    "boxSizing": "content-box",
                    "margin": 0,
                    "muiPrepared": true,
                    "position": "absolute",
                    "width": "100%",
                  }
                }
              />
              <hr
                aria-hidden="true"
                style={
                  Object {
                    "borderBottomStyle": "solid",
                    "borderBottomWidth": 2,
                    "borderColor": "#00bcd4",
                    "borderLeft": "none",
                    "borderRight": "none",
                    "borderTop": "none",
                    "bottom": 8,
                    "boxSizing": "content-box",
                    "margin": 0,
                    "muiPrepared": true,
                    "position": "absolute",
                    "transform": "scaleX(0)",
                    "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                    "width": "100%",
                  }
                }
              />
            </div>
          </div>
          <div>
            <button
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseOut={[Function]}
              onTouchStart={[Function]}
              style={
                Object {
                  "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
                  "background": "none",
                  "border": 10,
                  "boxSizing": "border-box",
                  "cursor": "pointer",
                  "display": "inline-block",
                  "fontFamily": "Roboto, sans-serif",
                  "fontSize": 0,
                  "fontWeight": "inherit",
                  "height": 48,
                  "margin": 0,
                  "muiPrepared": true,
                  "outline": "none",
                  "overflow": "visible",
                  "padding": 12,
                  "position": "relative",
                  "textDecoration": "none",
                  "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                  "verticalAlign": null,
                  "width": 48,
                }
              }
              tabIndex={0}
              type="button"
            >
              <div
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
              >
                <svg
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  style={
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                      "display": "inline-block",
                      "fill": "currentColor",
                      "height": 24,
                      "muiPrepared": true,
                      "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
                      "userSelect": "none",
                      "width": 24,
                    }
                  }
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
        <div
          style={
            Object {
              "muiPrepared": true,
              "padding": "8px 0px 8px 0px",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`StacksList when no stacks should match snapshot 1`] = `
<div
  style={
    Object {
      "overflow": "scroll",
    }
  }
>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "padding": 15,
        "paddingBottom": 30,
      }
    }
  />
</div>
`;
